### PR TITLE
Correct Interactive Block Behavior

### LINKED ISSUE #275: Items placed in the crafting table are shared globaly
### I've searched existing issues and couldn't find a duplicate.

- [X] I confirm this is not a duplicate.

### Operating System

Linux manjaro

### Server Software Version/Commit

public server

### What happened?

I placed Items in the /craft crafting table. Another player on the Server saw, what I put in and could "edit" the position of the Items.

### To Reproduce

/craft and put items in the crafting table. Another player should have acces to the items.

### Expected behavior

Crafting tables should stay independent for each player.

### PR DESCRIPTION
**Pull request by OfficialKris**
## Description
Fixes #275 and many other interactive block issues:
- All interactive blocks are per block now (for instance, chests and furnaces do not share containers globally)
- Crafting table containers are both unique per player AND per block
- Breaking interactive blocks removes all contents (can drop items once item drop is implemented)
- All active container users are forcibly closed when another `player` removes a block
- Removed artificial cap on container screen limit which kicked players after opening a container 15 times (we may not be handling container ids correctly)
- Added a `standard` set of container functions to reduce code duplication for interactive blocks

This PR requires #384 

## Testing
Two clients
- Place a crafting table, a furnace, and a chest
- Open each and place items in container
- Open crafting table again, notice no items are stored
- Open all other blocks and see items in both clients
- Open crafting table in both and notice containers are not shared

Internally, this change maintains the OpenContainer HashMap and a new AtomicU32 which tracks the current index for the generation of uuids. On large persistent servers, unique containers could be optimized to reduce memory usage.  Every user who opens a unique container will generate and cache a new data structure (think 10,000 users opening a crafting table, enderchest, etc). Future work could include separating unique containers (crafting table) and persistent unique containers (ender chest) into separate data structures (ex. stacks instead of maps).

Also, another optimization could be made to client interactive block interactions. Currently, this requires that the server iterate through the entirety of the OpenContainer HashMap. It may be better to create a sort of registry which maps block position to OpenContainerID in the long term (again large server optimizations). Otherwise, the algorithm is randomized brute force with a time complexity of O(n + m).

## Checklist
Things need to be done before this Pull Request can be merged.

- [x] Code is well-formatted and adheres to project style guidelines: `cargo fmt`
- [x] Code does not produce any clippy warnings: `cargo clippy`
- [x] All unit tests pass: `cargo test`
- [ ] I added new unit tests, so other people don't accidentally break my code by changing other parts of the codebase. [How?](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)

### CODE REVIEW

File: pumpkin/src/block/blocks/crafting_table.rs [hunk 1]
```diff
@@ -6,52 +6,112 @@ use async_trait::async_trait;
...
-                    Some(location),
-                    get_block("minecraft:crafting_table").cloned(),
-                );
-                open_containers.insert(1, open_container);
+                player
+                    .open_container
+                    .store(Some(id_to_use.try_into().unwrap()));
             }
         }
+        drop(open_containers);
```
Comment (DataM0del) [inline pumpkin/src/block/blocks/crafting_table.rs:114]: not sure what's the need for manually managing memory as if it was C

Comment (leobeg) [reply pumpkin/src/block/blocks/crafting_table.rs:114]: > not sure what's the need for manually managing memory as if it was C

I think it's not necassary here but it's more like freeing the guard then managing the memory. 
For example if you call an sub function that trys to lock the variable too you create an deadlock.
Which results in freezing

Comment (OfficialKris) [reply pumpkin/src/block/blocks/crafting_table.rs:114]: I think this is a draft and in progress :D I think I was trying to fix a different issue that I thought was from a deadlock. I will probably refactor this code

**Comment by Snowiiii** [comment]
When opening a Chest, the server freezes and i get a timeout

**Comment by OfficialKris** [comment]
Fixed, tried to optimize code like this was C.

**Comment by Snowiiii** [comment]
Okay works now, But i found a bug. If you shift click a item from a chest into your inventory and reopen the chest the item will be in the chest again

**Comment by OfficialKris** [comment]
I checked comparing with master and it seems shift clicking is just not supported yet. Shift clicking into containers does the same thing. I believe that should be a separate PR since it would be modifying other interaction container code.

**Comment by Snowiiii** [comment]
Waiting for @Bryntet review, but for some reason crafting does not work anymore :c

**Comment by Bryntet** [comment]
@Snowiiii IIRC crafting is currently broken on master too? or am I misremembering

**Comment by Snowiiii** [comment]
> @Snowiiii IIRC crafting is currently broken on master too? or am I misremembering

yep, its broken on master

**Comment by Snowiiii** [comment]
Merged! Thanks for your hard work on this, @OfficialKris :heart:

**Pull request merged**